File: app/src/main/java/com/limelight/ExternalDisplayManager.java
package com.limelight;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Presentation;
import android.content.Context;

import android.hardware.display.DisplayManager;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.util.TypedValue;
import android.view.Display;
import android.view.Gravity;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.limelight.binding.video.MediaCodecDecoderRenderer;
import com.limelight.nvstream.NvConnection;
import com.limelight.preferences.PreferenceConfiguration;
import com.limelight.ui.StreamView;
import com.limelight.LimeLog;
import com.limelight.utils.UiHelper;

/**
 * 外接显示器管理器
 * 负责管理外接显示器的检测、连接、断开和内容显示
 */
public class ExternalDisplayManager {
    
    private final Activity activity;
    private final PreferenceConfiguration prefConfig;
    private final NvConnection conn;
    private final MediaCodecDecoderRenderer decoderRenderer;
    private final String pcName;
    private final String appName;
    
    private DisplayManager displayManager;
    private Display externalDisplay;
    private boolean useExternalDisplay = false;
    private DisplayManager.DisplayListener displayListener;
    private ExternalDisplayPresentation externalPresentation;
    
    // 回调接口
    public interface ExternalDisplayCallback {
        void onExternalDisplayConnected(Display display);
        void onExternalDisplayDisconnected();
        void onStreamViewReady(StreamView streamView);
    }
    
    private ExternalDisplayCallback callback;
    
    public ExternalDisplayManager(Activity activity, PreferenceConfiguration prefConfig, 
                                 NvConnection conn, MediaCodecDecoderRenderer decoderRenderer,
                                 String pcName, String appName) {
        this.activity = activity;
        this.prefConfig = prefConfig;
        this.conn = conn;
        this.decoderRenderer = decoderRenderer;
        this.pcName = pcName;
        this.appName = appName;
    }
    
    public void setCallback(ExternalDisplayCallback callback) {
        this.callback = callback;
    }
    
    /**
     * 初始化外接显示器管理器
     */
    public void initialize() {
        // 初始化显示管理器
        displayManager = (DisplayManager) activity.getSystemService(Context.DISPLAY_SERVICE);
        
        // 设置显示器监听器
        setupDisplayListener();
        
        // 检查是否有外接显示器
        checkForExternalDisplay();
        
        // 如果有外接显示器，启动外接显示器演示，并降低内建屏幕亮度到30%
        if (useExternalDisplay) {
            Window window = activity.getWindow();
            if (window != null) {
                WindowManager.LayoutParams layoutParams = window.getAttributes();
                layoutParams.screenBrightness = 0.3f;
                window.setAttributes(layoutParams);
            }
            startExternalDisplayPresentation();
        }
    }
    
    /**
     * 清理资源
     */
    public void cleanup() {
        // 清理外接显示器演示
        if (externalPresentation != null) {
            externalPresentation.dismiss();
            externalPresentation = null;
        }
        
        // 取消注册显示器监听器
        if (displayListener != null && displayManager != null) {
            displayManager.unregisterDisplayListener(displayListener);
            displayListener = null;
        }
    }
    
    /**
     * 获取要使用的显示器
     */
    public Display getTargetDisplay() {
        if (useExternalDisplay && externalDisplay != null) {
            return externalDisplay;
        }
        return activity.getWindowManager().getDefaultDisplay();
    }
    
    /**
     * 检查是否正在使用外接显示器
     */
    public boolean isUsingExternalDisplay() {
        return useExternalDisplay && externalDisplay != null;
    }
    
    /**
     * 检查是否有外接显示器连接
     */
    public static boolean hasExternalDisplay(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            DisplayManager displayManager = (DisplayManager) context.getSystemService(Context.DISPLAY_SERVICE);
            if (displayManager != null) {
                Display[] displays = displayManager.getDisplays();
                for (Display display : displays) {
                    if (display.getDisplayId() != Display.DEFAULT_DISPLAY) {
                        return true;
                    }
                }
            }
        }
        return false;
    }
    
    /**
     * 设置显示器监听器
     */
    private void setupDisplayListener() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            displayListener = new DisplayManager.DisplayListener() {
                @Override
                public void onDisplayAdded(int displayId) {
                    LimeLog.info("Display added: " + displayId);
                    if (prefConfig.useExternalDisplay && displayId != Display.DEFAULT_DISPLAY) {
                        // 外接显示器已连接
                        checkForExternalDisplay();
                        if (useExternalDisplay) {
                            startExternalDisplayPresentation();
                        }
                    }
                }

                @Override
                public void onDisplayRemoved(int displayId) {
                    LimeLog.info("Display removed: " + displayId);
                    if (externalDisplay != null && displayId == externalDisplay.getDisplayId()) {
                        // 外接显示器已断开
                        if (externalPresentation != null) {
                            externalPresentation.dismiss();
                            externalPresentation = null;
                        }
                        externalDisplay = null;
                        useExternalDisplay = false;

                        // 显示主屏幕内容
                        View surfaceView = activity.findViewById(R.id.surfaceView);
                        if (surfaceView != null) {
                            surfaceView.setVisibility(View.VISIBLE);
                        }
                        Toast.makeText(activity, "外接显示器已断开，切换到主屏幕", Toast.LENGTH_SHORT).show();
                        
                        if (callback != null) {
                            callback.onExternalDisplayDisconnected();
                        }
                    }
                }

                @Override
                public void onDisplayChanged(int displayId) {
                    LimeLog.info("Display changed: " + displayId);
                }
            };

            displayManager.registerDisplayListener(displayListener, null);
        }
    }

    /**
     * 检查并配置外接显示器
     */
    private void checkForExternalDisplay() {
        // 如果用户没有启用外接显示器选项，直接返回
        if (!prefConfig.useExternalDisplay) {
            LimeLog.info("External display disabled by user preference");
            return;
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            Display[] displays = displayManager.getDisplays();

            // 查找外接显示器（不是主显示器）
            for (Display display : displays) {
                if (display.getDisplayId() != Display.DEFAULT_DISPLAY) {
                    externalDisplay = display;
                    useExternalDisplay = true;
                    LimeLog.info("Found external display: " + display.getName() +
                            " (ID: " + display.getDisplayId() + ")");
                    
                    if (callback != null) {
                        callback.onExternalDisplayConnected(display);
                    }
                    break;
                }
            }

            if (!useExternalDisplay) {
                LimeLog.info("No external display found, using default display");
            }
        }
    }

    /**
     * 将Activity移动到外接显示器
     */
    private void moveToExternalDisplay() {
        if (useExternalDisplay && externalDisplay != null &&
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {

            // 创建WindowManager.LayoutParams for external display
            WindowManager.LayoutParams params = activity.getWindow().getAttributes();
            params.preferredDisplayModeId = externalDisplay.getMode().getModeId();
            activity.getWindow().setAttributes(params);

            // 或者使用Presentation来在外接显示器上显示
            // 这需要重新设计Activity结构
        }
    }

    /**
     * 外接显示器演示类
     */
    private class ExternalDisplayPresentation extends Presentation {

        public ExternalDisplayPresentation(Context outerContext, Display display) {
            super(outerContext, display);
        }

        @Override
        protected void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);

            // 设置全屏
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
            getWindow().getDecorView().setSystemUiVisibility(
                    View.SYSTEM_UI_FLAG_LAYOUT_STABLE |
                            View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION |
                            View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);



            // 设置内容视图
            setContentView(R.layout.activity_game);

            // 初始化StreamView
            StreamView externalStreamView = findViewById(R.id.surfaceView);
            if (externalStreamView != null) {
                // 通知回调StreamView已准备就绪
                if (callback != null) {
                    callback.onStreamViewReady(externalStreamView);
                }
            }
        }

        @Override
        public void onDisplayRemoved() {
            super.onDisplayRemoved();
            // 外接显示器被移除时，关闭串流
            activity.finish();
        }
    }

    /**
     * 启动外接显示器演示
     */
    @SuppressLint({"ResourceAsColor", "SetTextI18n"})
    private void startExternalDisplayPresentation() {
        if (!(useExternalDisplay && externalDisplay != null && externalPresentation == null)) {
            return;
        }

        externalPresentation = new ExternalDisplayPresentation(activity, externalDisplay);
        externalPresentation.show();

        // 隐藏主Activity的内容
        View surfaceView = activity.findViewById(R.id.surfaceView);
        if (surfaceView != null) {
            surfaceView.setVisibility(View.GONE);
        }

        if (prefConfig.enablePerfOverlay) {
            // 创建电量显示TextView
            final TextView batteryTextView = new TextView(activity);
            batteryTextView.setGravity(Gravity.CENTER);
            batteryTextView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 48);
            batteryTextView.setTextColor(activity.getResources().getColor(R.color.scene_color_1));

            // 设置布局参数（居中显示）
            FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
                    FrameLayout.LayoutParams.WRAP_CONTENT,
                    FrameLayout.LayoutParams.WRAP_CONTENT
            );
            params.gravity = Gravity.CENTER;
            batteryTextView.setLayoutParams(params);

            // 添加到内建屏幕（主Activity）视图的中间
            FrameLayout rootView = activity.findViewById(android.R.id.content);
            if (rootView != null) {
                rootView.addView(batteryTextView);
            }

            // 创建定时更新任务
            final Handler handler = new Handler();
            final Runnable updateBatteryTask = new Runnable() {
                private final int[] gravityOptions = {
                    Gravity.CENTER,
                    Gravity.TOP | Gravity.CENTER_HORIZONTAL,
                    Gravity.BOTTOM | Gravity.CENTER_HORIZONTAL,
                    Gravity.CENTER_VERTICAL | Gravity.LEFT,
                    Gravity.CENTER_VERTICAL | Gravity.RIGHT,
                    Gravity.TOP | Gravity.LEFT,
                    Gravity.TOP | Gravity.RIGHT,
                    Gravity.BOTTOM | Gravity.LEFT,
                    Gravity.BOTTOM | Gravity.RIGHT
                };
                
                @Override
                public void run() {
                    // 更新电量显示
                    batteryTextView.setText(String.format("🔋 %d%%", UiHelper.getBatteryLevel(activity)));
                    
                    // 随机选择位置和参数以避免烧屏
                    int randomGravity = gravityOptions[(int) (Math.random() * gravityOptions.length)];
                    
                    // 随机生成边距参数（-200到200像素之间）
                    int randomMarginLeft = (int) (Math.random() * 401) - 200;
                    int randomMarginTop = (int) (Math.random() * 401) - 200;
                    int randomMarginRight = (int) (Math.random() * 401) - 200;
                    int randomMarginBottom = (int) (Math.random() * 401) - 200;
                    
                    FrameLayout.LayoutParams params = (FrameLayout.LayoutParams) batteryTextView.getLayoutParams();
                    params.gravity = randomGravity;
                    params.setMargins(randomMarginLeft, randomMarginTop, randomMarginRight, randomMarginBottom);
                    batteryTextView.setLayoutParams(params);
                    
                    // 每分钟更新一次
                    handler.postDelayed(this, 60000);
                }
            };

            // 立即执行首次更新并启动定时器
            updateBatteryTask.run();
        }

        Toast.makeText(activity, "串流已切换到外接显示器, 若某些外接设备不能正常横屏显示，请翻滚主机。", Toast.LENGTH_LONG).show();
    }
}

File: app/src/main/java/com/limelight/ExternalDisplayManager.kt
@file:Suppress("DEPRECATION")
package com.limelight

import android.app.Activity
import android.app.Presentation
import android.content.Context
import android.content.SharedPreferences
import android.hardware.display.DisplayManager
import android.os.Bundle
import android.view.Display
import android.view.View
import android.view.WindowManager
import android.widget.ImageView
import android.widget.Toast
import androidx.preference.PreferenceManager
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.limelight.preferences.BackgroundSource
import com.limelight.preferences.PreferenceConfiguration
import com.limelight.ui.StreamView
import java.io.File

/**
 * 外接显示器管理器
 * 负责管理外接显示器的检测、连接、断开和内容显示
 */
class ExternalDisplayManager(
    private val activity: Activity,
    private val prefConfig: PreferenceConfiguration,
    private val targetDisplayResolver: TargetDisplayResolver
) {
    private val displayManager =
        activity.getSystemService(Context.DISPLAY_SERVICE) as DisplayManager
    private var displayListener: DisplayManager.DisplayListener? = null
    private var externalPresentation: ExternalDisplayPresentation? = null
    private var dualScreenPresentation: DualScreenControlPresentation? = null
    private var idleBackgroundPresentation: IdleBackgroundPresentation? = null
    private var dualScreenControlsEnabled = true
    private var displayModeSelection: DisplayModeManager.DisplayModeSelection? = null
    private var backgroundPrefsListener: SharedPreferences.OnSharedPreferenceChangeListener? = null

    interface ExternalDisplayCallback {
        fun onExternalDisplayConnected(display: Display)
        fun onExternalDisplayDisconnected()
        fun onStreamViewReady(streamView: StreamView)
        fun onDualScreenControlPanelReady(
            rootView: View,
            streamDisplay: Display,
            controlDisplay: Display
        )
        fun onDualScreenDisconnected()
    }

    var callback: ExternalDisplayCallback? = null

    fun initialize(initialDisplayMode: DisplayModeManager.DisplayModeSelection? = null) {
        displayModeSelection = initialDisplayMode
        registerIdleBackgroundPreferenceListener()
        setupDisplayListener()
        reconcileDisplays()
    }

    fun cleanup() {
        dismissExternalPresentation()
        dismissDualScreenPresentation()
        dismissIdleBackgroundPresentation()
        unregisterIdleBackgroundPreferenceListener()

        displayListener?.let {
            displayManager.unregisterDisplayListener(it)
            displayListener = null
        }
    }

    fun getTargetDisplay(): Display {
        return targetDisplayResolver.currentDisplay()
    }

    fun isUsingExternalDisplay(): Boolean = targetDisplayResolver.isExternalDisplaySelected()

    fun setDualScreenControlsEnabled(enabled: Boolean) {
        dualScreenControlsEnabled = enabled
        if (!enabled) {
            val hadControlPresentation = dualScreenPresentation != null
            dismissDualScreenPresentation()
            if (hadControlPresentation) {
                callback?.onDualScreenDisconnected()
            }
            return
        }

        reconcileDisplays()
    }

    fun canRestoreDualScreenControls(): Boolean {
        if (dualScreenControlsEnabled || !prefConfig.useExternalDisplay) return false
        val streamDisplay = targetDisplayResolver.currentDisplay()
        return streamDisplay.displayId == Display.DEFAULT_DISPLAY &&
            targetDisplayResolver.controlDisplayFor(streamDisplay.displayId) != null
    }

    /**
     * Stores the mode selected for a specific display and applies it to the Presentation when
     * that display is rendered. The display id prevents a stale mode id from being used after a
     * hotplug event changes the target display.
     */
    fun updateDisplayMode(selection: DisplayModeManager.DisplayModeSelection) {
        displayModeSelection = selection

        if (externalPresentation?.isForDisplay(selection.displayId) == true) {
            externalPresentation?.window?.let { DisplayModeWindowApplier.apply(it, selection) }
        }
    }

    private fun setupDisplayListener() {
        val listener = object : DisplayManager.DisplayListener {
            override fun onDisplayAdded(displayId: Int) {
                LimeLog.info("Display added: $displayId")
                if (prefConfig.useExternalDisplay && displayId != Display.DEFAULT_DISPLAY) {
                    reconcileDisplays()
                }
            }

            override fun onDisplayRemoved(displayId: Int) {
                LimeLog.info("Display removed: $displayId")
                val wasIdleBackground =
                    idleBackgroundPresentation?.isForDisplay(displayId) == true
                if (wasIdleBackground) {
                    dismissIdleBackgroundPresentation()
                }

                val wasDualScreen = dualScreenPresentation?.isForDisplay(displayId) == true
                if (wasDualScreen) {
                    dismissDualScreenPresentation()
                    callback?.onDualScreenDisconnected()
                }

                val wasTargetDisplay = displayId != Display.DEFAULT_DISPLAY &&
                    targetDisplayResolver.onDisplayRemoved(displayId)
                if (wasTargetDisplay) {
                    dismissExternalPresentation()

                    if (callback != null) {
                        val surfaceView = activity.findViewById<View>(R.id.surfaceView)
                        surfaceView?.visibility = View.VISIBLE
                        Toast.makeText(activity, activity.getString(R.string.toast_external_display_disconnected), Toast.LENGTH_SHORT).show()

                        callback?.onExternalDisplayDisconnected()
                    }
                }

                if (wasIdleBackground || wasDualScreen || wasTargetDisplay) {
                    reconcileDisplays()
                }
            }

            override fun onDisplayChanged(displayId: Int) {
                LimeLog.info("Display changed: $displayId")
            }
        }

        displayListener = listener
        displayManager.registerDisplayListener(listener, null)
    }

    private fun dismissExternalPresentation() {
        externalPresentation?.dismiss()
        externalPresentation = null
    }

    private fun dismissDualScreenPresentation() {
        dualScreenPresentation?.dismiss()
        dualScreenPresentation = null
    }

    private fun dismissIdleBackgroundPresentation() {
        idleBackgroundPresentation?.clearBackground()
        idleBackgroundPresentation?.dismiss()
        idleBackgroundPresentation = null
    }

    private fun reconcileDisplays() {
        if (!prefConfig.useExternalDisplay) {
            LimeLog.info("Dual-screen and external display support disabled by user preference")
            targetDisplayResolver.resolve(false)
            dismissExternalPresentation()
            dismissDualScreenPresentation()
            dismissIdleBackgroundPresentation()
            return
        }

        val streamDisplay = targetDisplayResolver.resolve(true)
        if (callback == null) {
            val activityDisplayId = activity.windowManager.defaultDisplay.displayId
            val idleDisplay = displayManager.displays
                .sortedBy { it.displayId }
                .firstOrNull { it.displayId != activityDisplayId }
            if (idleDisplay == null) {
                LimeLog.info("No secondary display found for settings background")
                dismissIdleBackgroundPresentation()
                return
            }

            LimeLog.info(
                "Showing settings background on ${idleDisplay.name} " +
                    "(ID: ${idleDisplay.displayId})"
            )
            startIdleBackgroundPresentation(idleDisplay)
            return
        }

        dismissIdleBackgroundPresentation()
        val controlDisplay = targetDisplayResolver.controlDisplayFor(streamDisplay.displayId)
        if (controlDisplay == null) {
            LimeLog.info("No secondary display found, using default display")
            dismissExternalPresentation()
            dismissDualScreenPresentation()
            return
        }

        if (streamDisplay.displayId != Display.DEFAULT_DISPLAY) {
            LimeLog.info(
                "Using selected stream display: ${streamDisplay.name} " +
                    "(ID: ${streamDisplay.displayId}); controls on display ${controlDisplay.displayId}"
            )
            startExternalDisplayPresentation(streamDisplay)
            return
        }

        LimeLog.info(
            "Using selected stream display: ${streamDisplay.name} " +
                "(ID: ${streamDisplay.displayId}); controls on display ${controlDisplay.displayId}"
        )
        if (!dualScreenControlsEnabled) {
            LimeLog.info("Dual-screen controls hidden for the current streaming session")
            dismissDualScreenPresentation()
            return
        }
        startDualScreenPresentation(controlDisplay, streamDisplay)
    }

    private inner class ExternalDisplayPresentation(
        outerContext: Context,
        display: Display
    ) : Presentation(outerContext, display) {
        private val presentationDisplayId = display.displayId

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)

            window?.addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN)
            @Suppress("DEPRECATION")
            window?.decorView?.systemUiVisibility =
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                        View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION or
                        View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN

            val selection = displayModeSelection
            if (selection != null && selection.displayId == presentationDisplayId) {
                window?.let { DisplayModeWindowApplier.apply(it, selection) }
            }

            setContentView(R.layout.activity_game)

            val externalStreamView = findViewById<StreamView>(R.id.surfaceView)
            if (externalStreamView != null) {
                callback?.onStreamViewReady(externalStreamView)
            }
        }

        override fun onDisplayRemoved() {
            super.onDisplayRemoved()
            // DisplayManager.DisplayListener restores the main StreamView and hides the
            // lower-screen dashboard. Finishing here races that fallback path.
        }

        fun isForDisplay(displayId: Int): Boolean = presentationDisplayId == displayId
    }

    private inner class DualScreenControlPresentation(
        outerContext: Context,
        private val controlDisplay: Display,
        private val streamDisplay: Display
    ) : Presentation(outerContext, controlDisplay) {
        private val presentationDisplayId = controlDisplay.displayId

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)

            window?.addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN)
            @Suppress("DEPRECATION")
            window?.decorView?.systemUiVisibility =
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                    View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION or
                    View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or
                    View.SYSTEM_UI_FLAG_HIDE_NAVIGATION or
                    View.SYSTEM_UI_FLAG_FULLSCREEN or
                    View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY

            setContentView(R.layout.dual_screen_control_panel)
            findViewById<View>(R.id.dualScreenControlPanel)?.let { panelRoot ->
                callback?.onDualScreenControlPanelReady(
                    panelRoot,
                    streamDisplay,
                    controlDisplay
                )
            }
        }

        fun isForDisplay(displayId: Int): Boolean = presentationDisplayId == displayId
    }

    private inner class IdleBackgroundPresentation(
        outerContext: Context,
        display: Display
    ) : Presentation(outerContext, display) {
        private val presentationDisplayId = display.displayId

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)

            window?.setBackgroundDrawableResource(R.color.advance_setting_background)
            window?.addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN)
            @Suppress("DEPRECATION")
            window?.decorView?.systemUiVisibility =
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                    View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION or
                    View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or
                    View.SYSTEM_UI_FLAG_HIDE_NAVIGATION or
                    View.SYSTEM_UI_FLAG_FULLSCREEN or
                    View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY

            setContentView(R.layout.dual_screen_idle_background)
            refreshBackground()
        }

        fun refreshBackground() {
            val imageView = findViewById<ImageView>(R.id.dualScreenIdleBackgroundImage) ?: return
            Glide.with(activity).clear(imageView)
            imageView.setImageDrawable(null)

            val target = BackgroundSource.current(activity).resolveTarget(
                activity,
                resources.configuration.orientation
            ) ?: return
            val model: Any = if (target.startsWith("http")) target else File(target)
            Glide.with(activity)
                .load(model)
                .centerCrop()
                .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
                .into(imageView)
        }

        fun clearBackground() {
            findViewById<ImageView>(R.id.dualScreenIdleBackgroundImage)?.let { imageView ->
                Glide.with(activity).clear(imageView)
                imageView.setImageDrawable(null)
            }
        }

        fun isForDisplay(displayId: Int): Boolean = presentationDisplayId == displayId
    }

    private fun startExternalDisplayPresentation(display: Display) {
        if (!isUsingExternalDisplay() || externalPresentation?.isForDisplay(display.displayId) == true) {
            return
        }

        if (dualScreenPresentation != null) {
            dismissDualScreenPresentation()
            callback?.onDualScreenDisconnected()
        }

        callback?.onExternalDisplayConnected(display)
        externalPresentation = ExternalDisplayPresentation(activity, display)
        externalPresentation?.show()

        val surfaceView = activity.findViewById<View>(R.id.surfaceView)
        surfaceView?.visibility = View.GONE

        Toast.makeText(activity, activity.getString(R.string.toast_switched_to_external_display), Toast.LENGTH_LONG).show()
    }

    private fun startIdleBackgroundPresentation(display: Display) {
        if (idleBackgroundPresentation?.isForDisplay(display.displayId) == true) {
            idleBackgroundPresentation?.refreshBackground()
            return
        }

        dismissExternalPresentation()
        dismissDualScreenPresentation()
        dismissIdleBackgroundPresentation()
        idleBackgroundPresentation = IdleBackgroundPresentation(activity, display)
        idleBackgroundPresentation?.show()
    }

    private fun startDualScreenPresentation(controlDisplay: Display, streamDisplay: Display) {
        if (externalPresentation != null ||
            dualScreenPresentation?.isForDisplay(controlDisplay.displayId) == true
        ) {
            return
        }

        dualScreenPresentation = DualScreenControlPresentation(
            activity,
            controlDisplay,
            streamDisplay
        )
        dualScreenPresentation?.show()
        Toast.makeText(
            activity,
            activity.getString(R.string.toast_dual_screen_controls_ready),
            Toast.LENGTH_LONG
        ).show()
    }

    private fun registerIdleBackgroundPreferenceListener() {
        if (callback != null || backgroundPrefsListener != null) return

        val listener = SharedPreferences.OnSharedPreferenceChangeListener { _, key ->
            if (key == BackgroundSource.KEY_SOURCE ||
                key == BackgroundSource.KEY_API_URL ||
                key == BackgroundSource.KEY_LOCAL_PATH
            ) {
                activity.runOnUiThread {
                    idleBackgroundPresentation?.refreshBackground()
                }
            }
        }
        backgroundPrefsListener = listener
        PreferenceManager.getDefaultSharedPreferences(activity)
            .registerOnSharedPreferenceChangeListener(listener)
    }

    private fun unregisterIdleBackgroundPreferenceListener() {
        backgroundPrefsListener?.let { listener ->
            PreferenceManager.getDefaultSharedPreferences(activity)
                .unregisterOnSharedPreferenceChangeListener(listener)
        }
        backgroundPrefsListener = null
    }

    companion object {
        fun hasExternalDisplay(context: Context): Boolean {
            val displayManager = context.getSystemService(Context.DISPLAY_SERVICE) as? DisplayManager
            if (displayManager != null) {
                for (display in displayManager.displays) {
                    if (display.displayId != Display.DEFAULT_DISPLAY) {
                        return true
                    }
                }
            }
            return false
        }
    }
}
